### ISSUE TITLE
Converge on a single instrumentation codebase?

### ISSUE THREAD

**Issue by gotwarlost** [istanbuljs/istanbul-lib-instrument#4]
/to @dtinth @bcoe 
/cc @davglass

Hi guys,

I started to work on istanbul v1 after a 6 month hiatus of totally ignoring it. I have, like, 3 weeks 
to get a reasonable "v1" out and I'm also looking for an exit strategy in terms of handing modules over
to people who can actually work on it and make it better. Nowadays, it is getting harder and harder for
me to work in this.

In the process, I discovered the new babel plugin-based instrumenter by @dtinth and there are a lot of 
good ideas in it, not to mention the simplicity of the code since Babel handles the heavy lifting of 
the AST transformation. 

The thing that bothered me, though, is that it is yet another variant of code coverage transformations and has different semantics from istanbul's instrumenter (notably the inability to ignore code for instrumentation and a bunch of other smaller differences). Now there is istanbul 0.x, istanbul 1.x-alpha and the `__coverage__` babel plugin each of which do subtly to wildly different things and cause friction for the end user who wants to move between ES5/ Typescript/ ES6.

So, @dtinth I see your "shamelessly copied from istanbul" code and raise you my version of shameless copies from your code. 

The `babel` branch in this repo is a sketch of an implementation of instrumentation as a AST babel visitor. It implements "ignores" and steals/ incorporates all the good ideas from the `__coverage__` implementation. (There may be some differences that I need to reconcile in a second pass based on what you guys think).

The istanbul instrumentation API is now witten in ES6, and adapted to use Babel under the covers. The programVisitor that is exposed can be be used by a babel plugin that implements source file filters and calls the visitor only when code needs to be instrumented. This, IMO, gives us some semblance of consistency for ES5/ ES6 code coverage.

I'll let you guys check out the code and play with it. All my current tests pass, although es6 tests are pretty light.

I would love to get some early feedback on whether this looks like a promising approach that can be taken to completion. 

Also, we can open up this repo for commit access to more folks so you don't have to deal with the 
"@gotwarlost is unresponsive" issue :)

Please add more people to the thread as you see fit. I also have some other ideas re: the CLI (TLDR; make `nyc` the default CLI for istanbul) that we can discuss as we go.

Thanks,
Krishnan

**Comment by dtinth** [comment]
This is really cool. The `babel-plugin-__coverage__` project also has some tests regarding ES6, so please feel free to copy them. Your version of the code also looks very clean.

The way `babel-plugin-__coverage__` treats statements is quite unconventional, and if it wouldn’t become utterly inefficient, it would have treated every kind of expression as a statement. For example, in my “ideal” version (that doesn’t exist) this line of code would be 5 statements.

```js
    doSomethingWith(objectA, objectB, objectC)
//  <-------------> S1
//                  <-----> S2
//                           <-----> S3
//                                    <-----> S4
//  <----------------------------------------> S5
```

This is because `objectB` may be undeclared, which causes a ReferenceError, and that means S4 and S5 will not be evaluated.

I’m not sure, but I have this speculation that what I’ve been doing in `babel-plugin-__coverage__` may be more adequately covered using “branch coverage”, and if it makes CodeCov turn yellow in report, then I’m also fine with that. For example, an AssignmentPattern can be a “branch” rather than a “statement” instead.

**Comment by gotwarlost** [comment]
@dtinth - thanks for the prompt reply and encouragement. Let me take a stab at implementing missing es6 tests and a few other changes.

Re, your `doSomethingWith` example, I'm not sure that we should worry about getting coverage numbers right on reference errors and such - I think it is OK to say that coverage is accurate only if all tests pass. Otherwise we might end up with too much information and counters that cannot be intuitively explained. From the very beginning my mantra was not to get code coverage "right" in some technical sense of the word but be "good enough" in a way that helps the user with an actionable item that the user can choose (or not) to deal with.

I agree that assignment patterns and `for` statement update expressions are better represented as branches. Let me go down this path and see if this can be done. This might also need a couple of fixes to the logic of how branches are colored in the HTML report.

The idea behind my "spec" tests are sound in the sense that these tests have helped me completely change the internals of the implementation (from uglify to esprima to babel) and have some confidence that the changes are correct. However, the implementation is quite horrible and I plan to fix this as well.
I'll also add more spec tests for es6 features.

Let me work on this for a day or two and I'll show you a more refined product and maybe send a PR for the `__coverage__` plugin to use it. I'll also do a PR for this repo so I can get review comments from you and others. My experience with writing ES6 code is a whopping 2 days and I'm sure the code that I write will not be idiomatic ES6.

Thanks again for the quick feedback.

**Comment by bcoe** [comment]
@gotwarlost @dtinth this seems like a wonderful idea, and I'm excited to keep working with you to help make `nyc` extensible, so that folks can plugin whatever instrumenters they see fit -- I imagine we'll move towards this being the default one.

I'll dive into this branch on weeknights this week, and will carve out some time on the weekend for hacking as well.

**Comment by graingert** [comment]
@gotwarlost have a look at https://github.com/douglasduteil/isparta also

**Comment by graingert** [comment]
Here's a link directly to that __coverage__ project https://github.com/dtinth/babel-plugin-__coverage__

**Comment by graingert** [comment]
Ping @deepsweet @douglasduteil

**Comment by Kflash** [comment]
@gotwarlost Is the plan that Babel need to be a dependency for the new version? You also have a ES2015 compiler - Buble. With current version it works, but will not work if Babel is a required dependency.

**Comment by graingert** [comment]
@Kflash babel-plugin-\_\_coverage\_\_ only needs:

```js
"babel-helper-function-name": "^6.5.0",
"babel-template": "^6.8.0",
"test-exclude": "^1.1.0"
```

all it does is add the coverage object to the correct location in the code.

**Comment by Kflash** [comment]
@graingert I know, but I got a hard time getting it working with Buble. In fact I couldn't. Isparta and current Istanbul is the only version that works it seems. Istanbul-alpha-2 too give issues.

@gotwarlost Another strange thing in this is the CLI isn't working without issues on a Windows computer, but works smooth on a Mac. Isparta was mentioned here, and that CLI works without any issues with Windows. Maybe study that code?

**Comment by gotwarlost** [comment]
@Kflash - there are 2 ways to use the library.

1. As a babel plugin that uses the exposed `programVisitor` to do the instrumentation
2. By calling the instrumentation API which is (mostly) backwards-compatible to the current istanbul API. It uses babel libraries under the covers but does not expose its internals outside.

It should work based on all my unit tests that continue to pass for all node versions.

See: https://github.com/istanbuljs/istanbul-lib-instrument/pull/5

Hope this clarifies.

**Comment by gotwarlost** [comment]
Resolved via #5

**Comment by Kflash** [comment]
@gotwarlost Can you look at this command, and let me know why it doesn't generate any report? Tests run just fine: https://github.com/jeibo/workflow/blob/master/scripts/coveralls.sh
